Implementation of `mockCall` cheat code (#318)
runtimeverification/kontrol: src/kontrol/kdist/cheatcodes.md

          <addressSet> .Set </addressSet>
          <storageSlotSet> .Set </storageSlotSet>
        </whitelist>

>         <mockCalls>

Reviewer: @lisandrasilva  One other thing I forgot to point out. Above this configuration there is a short description of each cell and its role. Could you update that as well?

Author: Sure :)

Reviewer: Thank you!

      [owise]
```

```k
    rule [foundry.call.mockCall]:
         <k> #call_foundry SELECTOR ARGS
          => #loadAccount #asWord(#range(ARGS, 0, 32)) 
          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
         ...
         </k> 
      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )

    //rule [foundry.call.mockCallWithMsgValue]:
    //     <k> #call_foundry SELECTOR ARGS
    //        => #loadAccount #asWord(#range(ARGS, 0, 32))
    //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
    //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )
```

```k
    rule [foundry.set.mockCall]:
         <k> #next [ _OP:CallOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
         <localMem> LM </localMem>

>          <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>

Reviewer: To avoid warnings, you can replace unused args with `_`, or prefix their name with underscore.
```suggestion
         <wordStack> _ : ACCTTO : _ : ARGSTART : _ : RETSTART : RETWIDTH : _WS </wordStack>
```

         <mockCall>
           <mockAddress> ACCTTO </mockAddress>
           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
         </mockCall>
         requires #range(LM, ARGSTART, lengthBytes(CALLDATA)) ==K CALLDATA
      [priority(30)]

      rule [foundry.set.mockCall2]:
         <k> #next [ _OP:CallSixOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
         <localMem> LM </localMem>

>          <wordStack> _ : ACCTTO : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>

Reviewer: ```suggestion
         <wordStack> _ : ACCTTO : ARGSTART : _ : RETSTART : RETWIDTH : _WS </wordStack>
```

      [owise]
```

> 

Reviewer: Please add a description of the cheat code, including the current limitations that it has.

Author: I forgot to do that. Thanks!

```k
    rule [foundry.call.mockCall]:
         <k> #call_foundry SELECTOR ARGS
          => #loadAccount #asWord(#range(ARGS, 0, 32))
          ~> #etchAccountIfEmpty #asWord(#range(ARGS, 0, 32))
          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
         ...
         </k> 
      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )

> 
>     //rule [foundry.call.mockCallWithMsgValue]:
>     //     <k> #call_foundry SELECTOR ARGS
>     //        => #loadAccount #asWord(#range(ARGS, 0, 32))
>     //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
>     //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )

Reviewer: Please remove this commented code.
```suggestion
```

Author: Done!

      [owise]
```

```k
    rule [foundry.call.mockCall]:
         <k> #call_foundry SELECTOR ARGS
          => #loadAccount #asWord(#range(ARGS, 0, 32)) 
          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
         ...
         </k> 
      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )

    //rule [foundry.call.mockCallWithMsgValue]:
    //     <k> #call_foundry SELECTOR ARGS
    //        => #loadAccount #asWord(#range(ARGS, 0, 32))
    //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
    //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )
```

```k

>     rule [foundry.set.mockCall]:
>          <k> #next [ _OP:CallOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
>          <localMem> LM </localMem>
>          <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
>          <mockCall>
>            <mockAddress> ACCTTO </mockAddress>
>            <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
>          </mockCall>
>          requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
>       [priority(30)]

Reviewer: I think that here you could have something like this instead. Basically saying that if you have a mocked value for a given address, skip executing the `CALL` opcode and just set the mocked value in the local memory and output cell.
```suggestion
    rule [foundry.set.mockCall]:
         <k> #next [ _OP:CallOp ] =>  #setLocalMem RETSTART RETWIDTH RETURNDATA  ... </k>
         <localMem> LM </localMem>
         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
         <mockCall>
           <mockAddress> ACCTTO </mockAddress>
           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
         </mockCall>
         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
      [priority(30)]
```

Author: That was my first approach, but then I noticed (and tested) that actually if the function called changes the state of the contract, those changes are made, which means that the function is executed, but instead, it returns the value provided in the `mockCall`.

Reviewer: I see. Then you can leave them as they are. :+1:

      rule [foundry.set.mockCall2]:
         <k> #next [ _OP:CallSixOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
         <localMem> LM </localMem>
         <wordStack> _ : ACCTTO : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
         <mockCall>
           <mockAddress> ACCTTO </mockAddress>
           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
         </mockCall>
         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
      [priority(30)]
```

```k
    syntax KItem ::= "#updateOutput" Bytes [klabel("foundry_updateRevertOutput")]

>  // -------------------------------------------------------------------------------------

Reviewer: ```suggestion
 // -----------------------------------------------------------------------------
```

Reviewer: Also, where is this function used?

Author: It's not being used anymore. I will delete it.

      [owise]
```

Mock calls
----------

#### `mockCall` - Mocks all calls to an address where if the call data either strictly or loosely matches data and returns retdata.

```
function mockCall(address where, bytes calldata data, bytes calldata retdata) external;
```

`foundry.call.mockCall` will match when the `mockCall` cheat code function is called.

> This rule then takes the `address` value from the function call data and etches a single byte into the account code, in case it is empty.

Reviewer 2: It's a very minor thing, but I'd recommend to write "calldata" in one word to follow the Ethereum convention (here and below).

```suggestion
This rule then takes the `address` value from the function calldata and etches a single byte into the account code, in case it is empty.
```

> The rule also takes the bytes `CALLDATA` and the bytes `RETURNDATA` from the function call data and forwards them, together with the address to be mocked, and forwards them to the `setMockCall` production.

Reviewer 2: "Forwards them" is repeated twice:

```suggestion
The rule also takes the bytes `CALLDATA` and the bytes `RETURNDATA` from the function calldata and forwards them, together with the address to be mocked, to the `setMockCall` production.
```

The `setMockCall` production will update the configuration in order to store the information of the mock call.

The current implementation of the `mockCall` cheatcode has some limitations:
- It does not work if there are multiple mock calls with common prefixes for the same address - see test `MockCallTestFoundry.testMockCallMultiplePartialMatch`

```k
    rule [foundry.call.mockCall]:
         <k> #call_foundry SELECTOR ARGS
          => #loadAccount #asWord(#range(ARGS, 0, 32))
          ~> #etchAccountIfEmpty #asWord(#range(ARGS, 0, 32))
          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
         ...
         </k> 
      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )
```

> We use the `#next[OP]` to identify OpCodes that represent function calls. If there is `<mockCall>` which `<mockAddress>`

Reviewer 2: Shall this sentence be one-line as well? I think the wording here can also be a bit more streamlined, e.g., 
```suggestion
We use `#next[OP]` to identify OpCodes that represent function calls. If there is `<mockCall>`, for which `<mockAddress>` matches the `ACCTTO` and `<mockValues>` has a key `CALLDATA` that matches some prefix of the function calldata, then the `#execMockCall` will replace the function execution and update the output with the `RETURNDATA`.
```

      [owise]
```

Mock calls
----------

> #### `mockCall` - Mocks all calls to an address where if the call data either strictly or loosely

Reviewer 2: IIUC, the convention is to have one line per sentence in Markdown files (@anvacaru please correct me if I'm wrong!).

             <acctID> ACCT </acctID>
             <code> CODE => #bufStrict(1,0) </code>
             ...
           </account>
           ...
         </accounts>
      requires lengthBytes(CODE) ==Int 0
    rule <k> #etchAccountIfEmpty _ => . ... </k> [owise]
```

- `#setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN` will update the `<mockcalls>` mapping for the given account.

```k
    syntax KItem ::= "#setMockCall" Account Bytes Bytes [klabel(foundry_setMockCall)]
 // ---------------------------------------------------------------------------------
    rule <k> #setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN => . ... </k>
         <mockCall>
            <mockAddress> MOCKADDRESS </mockAddress>
            <mockValues>  MOCKVALUES => MOCKVALUES [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
         </mockCall>

   rule <k> #setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN => . ... </k>
         <mockCalls>
           ( .Bag
            => <mockCall>
                  <mockAddress> MOCKADDRESS </mockAddress>
                  <mockValues> .Map [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
               </mockCall>
           )
           ...
         </mockCalls>
```

```k

>     syntax KItem ::= "#clearMockCall" Int Int Bytes [klabel(foundry_clearMockCall)]
>  // -------------------------------------------------------------------------

Reviewer: Formatting.
```suggestion
    syntax KItem ::= "#clearMockCall" Int Int Bytes [klabel(foundry_clearMockCall)]
 // -------------------------------------------------------------------------------
```

    rule <k> #clearMockCall RETSTART RETWIDTH RETURNDATA => #setLocalMem RETSTART RETWIDTH RETURNDATA ... </k>
         <statusCode> _ => EVMC_SUCCESS </statusCode>
         <wordStack> _ : WS => 1 : WS </wordStack>
```

> 

Reviewer: Formatting.
```suggestion
```

        </whitelist>
```

- `#setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN` will update the `<mockcalls>` mapping for the given account.

```k
    syntax KItem ::= "#setMockCall" Account Bytes Bytes [klabel(foundry_setMockCall)]

>  // ------------------------------------------------------------------------

Reviewer: Formatting.
```suggestion
 // --------------------------------------------------------------------------------
```

        </whitelist>
```

> - `#etchAccountIfEmpty Account` - sets an Account code to a single byte '0u8' if the account is empty to circumvent the `extcodesize` check that Solidity might perform [source](https://github.com/foundry-rs/foundry/blob/b78289a0bc9df6e35624c632396e16f27d4ccb3f/crates/cheatcodes/src/evm/mock.rs#L54).

Reviewer 2: Sorry, I forgot to include it in the previous review. You can probably put the link to Foundry documentation in brackets as `(source)`, so it's more obvious that it's not part of the sentence :) 

```suggestion
- `#etchAccountIfEmpty Account` - sets an Account code to a single byte '0u8' if the account is empty to circumvent the `extcodesize` check that Solidity might perform ([source](https://github.com/foundry-rs/foundry/blob/b78289a0bc9df6e35624c632396e16f27d4ccb3f/crates/cheatcodes/src/evm/mock.rs#L54)).
```

             <acctID> ACCT </acctID>
             <code> CODE => #bufStrict(1,0) </code>
             ...
           </account>
           ...
         </accounts>
      requires lengthBytes(CODE) ==Int 0
    rule <k> #etchAccountIfEmpty _ => . ... </k> [owise]
```

- `#setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN` will update the `<mockcalls>` mapping for the given account.

```k
    syntax KItem ::= "#setMockCall" Account Bytes Bytes [klabel(foundry_setMockCall)]
 // ---------------------------------------------------------------------------------
    rule <k> #setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN => . ... </k>
         <mockCall>
            <mockAddress> MOCKADDRESS </mockAddress>
            <mockValues>  MOCKVALUES => MOCKVALUES [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
         </mockCall>

   rule <k> #setMockCall MOCKADDRESS MOCKCALLDATA MOCKRETURN => . ... </k>
         <mockCalls>
           ( .Bag
            => <mockCall>
                  <mockAddress> MOCKADDRESS </mockAddress>
                  <mockValues> .Map [ MOCKCALLDATA <- MOCKRETURN ] </mockValues>
               </mockCall>
           )
           ...
         </mockCalls>
```

> - `#execMockCall` will update the output of the function call with `RETURNDATA` using `#setLocalMem` and in case the function

Reviewer 2: I'd split this into two sentences, e.g., something like
```suggestion
- `#execMockCall` will update the output of the function call with `RETURNDATA` using `#setLocalMem`.
In case the function did not end with `EVMC_SUCCESS` it will update the status code to `EVMC_SUCCESS`. 
```
Or you can just add a comma between two clauses.

Author: Thanks for all the suggestions! I just addressed them.
